dev/AppNotifications/NotificationDevicesData.cpp

```diff
@@ -0,0 +1,41 @@
+// Copyright (c) Microsoft Corporation and Contributors.
+// Licensed under the MIT License.
+
+#include "pch.h"
+
+#include "AppNotificationConferencingConfig.h"
+#include "NotificationDevicesData.h"
+#include "AppNotificationUtility.h"
+
+using namespace Microsoft::Windows::AppNotifications::Helpers;
+
+NotificationDevicesData::NotificationDevicesData(winrt::AppNotificationConferencingConfig const& conferencingConfig)
+    : m_devicesData(conferencingConfig)
+{
+}
+
+STDMETHODIMP NotificationDevicesData::get_CameraDeviceId(_Out_ HSTRING* value) noexcept try
+{
+    *value = safe_make_unique_hstring(m_devicesData.VideoDeviceId().c_str()).release();
+    return S_OK;
+}
+CATCH_RETURN()
+
+STDMETHODIMP NotificationDevicesData::get_MicrophoneDeviceId(_Out_ HSTRING* value) noexcept try
+{
+    *value = safe_make_unique_hstring(m_devicesData.AudioInputDeviceId().c_str()).release();
+    return S_OK;
+}
+CATCH_RETURN() 
+
+
+STDMETHODIMP NotificationDevicesData::get_SpeakerDeviceId(_Out_ HSTRING* value) noexcept try
+{
+    *value = safe_make_unique_hstring(m_devicesData.AudioOutputDeviceId().c_str()).release();
```

Comment:
See https://learn.microsoft.com/en-us/windows/uwp/cpp-and-winrt-apis/interop-winrt-abi#interoperating-with-the-abis-hstring - consider `winrt::copy_to_abi(m_devicesData.AudioOutputDeviceId(), reinterpret_cast<void*&>(value))`

Reply:
Ok.. winrt::copy_to_abi, avoids unnecessary memory allocations (HString*) from safe_make_unique_hstring .
Thank you, Will change it.

Reply:
modified in the recent commit.